app/services/payment_provider_customers/adyen_service.rb

```diff
@@ -0,0 +1,125 @@
+# frozen_string_literal: true
+
+module PaymentProviderCustomers
+  class AdyenService < BaseService
+    def initialize(adyen_customer = nil)
+      @adyen_customer = adyen_customer
+
+      super(nil)
+    end
+
+    def create
+      result.adyen_customer = adyen_customer
+      return result if adyen_customer.provider_customer_id?
+
+      adyen_result = generate_checkout_url
+
+      result.checkout_url = adyen_result.response["url"]
+      result
+    end
+
+    def generate_checkout_url
+      res = client.checkout.payment_links_api.payment_links(payment_link_params)
+
+      SendWebhookJob.perform_later(
+        'customer.checkout_url_generated',
+        customer,
+        checkout_url: res.response["url"]
+      )
+
+      return res
+    rescue Adyen::AdyenError => e
+      deliver_error_webhook(e)
+
+      raise
+    end
+
+    def preauthorise(organization, event)
+      shopper_reference = shopper_reference_from_event(event)
+      payment_method_id = event.dig("additionalData", "recurring.recurringDetailReference")
+
+      @adyen_customer = PaymentProviderCustomers::AdyenCustomer.
+        joins(:customer).
+        where(customers: { external_id: shopper_reference, organization_id: organization.id }).
+        first
+
+      return handle_missing_customer(shopper_reference) unless adyen_customer
+
+      adyen_customer.update!(payment_method_id:, provider_customer_id: shopper_reference)
+
+      if organization.webhook_url?
+        SendWebhookJob.perform_later('customer.payment_provider_created', customer)
+      end
+
+      result.adyen_customer = adyen_customer
+      result
+    end
+
+    private
+
+    attr_accessor :adyen_customer
+
+    delegate :customer, to: :adyen_customer
+
+    def organization
+      @organization ||= customer.organization
+    end
+
+    def adyen_payment_provider
+      @adyen_payment_provider || organization.adyen_payment_provider
+    end
+
+    def client
+      @client ||= Adyen::Client.new(
+        api_key: adyen_payment_provider.api_key,
+        env: adyen_payment_provider.environment,
+        live_url_prefix: adyen_payment_provider.live_prefix
+      )
+    end
+
+    def shopper_reference_from_event(event)
+      event.dig("additionalData", "shopperReference") ||
```

Comment:
Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols. (https://rubystyle.guide#consistent-string-literals)